src/System.IO.Ports/src/System/IO/Ports/SerialPort.Uap.cs

```diff
@@ -10,8 +10,39 @@ public partial class SerialPort : Component
     {
         public static string[] GetPortNames()
         {
-            // See https://github.com/dotnet/corefx/issues/20588 for more information.
-            throw new PlatformNotSupportedException(System.SR.PlatformNotSupported_SerialPort_GetPortNames);
+            Span<uint> portNumbers = stackalloc uint[16];
+            uint portNumbersFound;
+            int error;
+
+            try
+            {
+                error = Interop.mincore.GetCommPorts(portNumbers, out portNumbersFound);
+            }
+            catch (Exception e) when (e is EntryPointNotFoundException || e is DllNotFoundException)
+            {
+                throw new PlatformNotSupportedException(System.SR.PlatformNotSupported_SerialPort_GetPortNames);
+            }
+
+            if (error == Interop.Errors.ERROR_MORE_DATA)
+            {
+                portNumbers = portNumbersFound <= 64 ?
+                    stackalloc uint[(int)portNumbersFound] :
+                    new uint[portNumbersFound];
+                error = Interop.mincore.GetCommPorts(portNumbers, out portNumbersFound);
+            }
+
+            if (error != Interop.Errors.ERROR_SUCCESS)
+            {
+                throw new Win32Exception(error);
+            }
+
+            var portNames = new string[portNumbersFound];
+            for (int i = 0; i < portNumbersFound; ++i)
+            {
+                portNames[i] = "COM" + portNumbers[i].ToString();
```

Comment:
Nit: You'd be better off doing $"COM{portNumbers[i]}". It will save an unnecessary allocation for each loop and looks a little cleaner.

Reply:
I am under the impression that the using interpolation will have a call to string.Format, boxing on the uint, and inside the string.Format a call to ToString() - plus likely a params array in some point of the stack. In the current format will be string.Concat and the ToString() so we saved the boxing and used string.Concat that is typically faster than string.Format. Is this some new optimization after 2.1 that I'm not aware? Not that it matters much in this case.

Reply:
Interesting, interpolation with string uses string.Concat but without the ToString() there is a On 2.1 here is what I get:

For `portNames[i] = "COM" + portNumbers[i].ToString();` and `portNames[i] = $"COM{portNumbers[i].ToString()}";`:
```
  IL_0094:  ldstr      "COM"
  IL_0099:  ldloca.s   V_0
  IL_009b:  ldloc.s    V_6
  IL_009d:  call       instance !0& valuetype [System.Runtime]System.Span`1<uint32>::get_Item(int32)
  IL_00a2:  call       instance string [System.Runtime]System.UInt32::ToString()
  IL_00a7:  call       string [System.Runtime]System.String::Concat(string, string)
  IL_00ac:  stelem.ref
```

For `portNames[i] = $"COM{portNumbers[i]}";`:
```
  IL_0094:  ldstr      "COM{0}"
  IL_0099:  ldloca.s   V_0
  IL_009b:  ldloc.s    V_6
  IL_009d:  call       instance !0& valuetype [System.Runtime]System.Span`1<uint32>::get_Item(int32)
  IL_00a2:  ldind.u4
  IL_00a3:  box        [System.Runtime]System.UInt32
  IL_00a8:  call       string [System.Runtime]System.String::Format(string, object)
  IL_00ad:  stelem.ref
```

Reply:
OR maybe you could use string.Create with SpanAction. https://github.com/dotnet/coreclr/blob/master/src/System.Private.CoreLib/shared/System/String.cs#L346

https://github.com/dotnet/corefx/pull/30528

Reply:
Hmm, forgot about the box. It would be interesting to benchmark differences. We have ISpanFormattable now that avoids string creation inside String::Format with int, uint, etc.

I don't care deeply here, but if we find they are roughly equivalent we may want to encourage the `$` syntax in the future. I'm investigating improving formatting methods on StreamWriter etc. and the possibility of getting the compiler to generate calls directly to the format overloads (and avoiding everything but the boxing).

Reply:
Yeah, it is really not that important since this API should be typically called only once from any app. Anyway I did measure and as usual benchmark/profile has some interesting info: the interpolation you suggested in the end has less allocations, however, it is the slowest by far. I quickly put together code using string.Create, I can't vouch it for correctness but should be on the ballpark of the required work, and it also surprises: great perf but not so great regarding allocation. I will keep the code as it is. Here are my results:

```
// * Summary *

BenchmarkDotNet=v0.11.0, OS=Windows 10.0.17134.167 (1803/April2018Update/Redstone4)
Intel Core i7-6700 CPU 3.40GHz (Skylake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=2.1.400-preview-009171
  [Host]     : .NET Core 2.1.2 (CoreCLR 4.6.26628.05, CoreFX 4.6.26629.01), 64bit RyuJIT
  DefaultJob : .NET Core 2.1.2 (CoreCLR 4.6.26628.05, CoreFX 4.6.26629.01), 64bit RyuJIT


                               Method |       Mean |     Error |    StdDev | Scaled | ScaledSD |  Gen 0 | Allocated |
------------------------------------- |-----------:|----------:|----------:|-------:|---------:|-------:|----------:|
                          AddOperator |   904.1 ns | 17.761 ns | 16.614 ns |   1.19 |     0.02 | 0.4005 |   1.65 KB |
              AddOperatorPlusToString |   758.6 ns |  7.396 ns |  6.556 ns |   1.00 |     0.00 | 0.3099 |   1.27 KB |
                        Interpolation | 2,044.9 ns | 47.806 ns | 55.053 ns |   2.70 |     0.07 | 0.2785 |   1.15 KB |
            InterpolationPlusToString |   768.1 ns | 17.538 ns | 16.405 ns |   1.01 |     0.02 | 0.3099 |   1.27 KB |
           StringCreatePlusSpanAction |   820.4 ns | 15.409 ns | 16.487 ns |   1.08 |     0.02 | 0.4320 |   1.77 KB |
 StringCreatePlusSpanActionWithLambda |   754.0 ns | 14.552 ns | 13.612 ns |   0.99 |     0.02 | 0.5236 |   2.15 KB |
```

Reply:
> Yeah, it is really not that important since this API should be typically called only once from any app.

Right. You should be measuring code size and picking up the one that has smallest code.